consensus/src/quorum_store/utils.rs

```diff
@@ -305,24 +335,59 @@ impl ProofQueue {
     ) {
         let start = Instant::now();
         for (batch_info, txn_summaries) in batch_summaries {
-            let batch_key = BatchKey::from_info(&batch_info);
-            if self
-                .batch_to_txn_summaries
-                .insert(batch_key, txn_summaries.clone())
-                .is_none()
+            // As txn_summary_num_occurrences count the number of times the txn_summary occurred in the
+            // remaining proofs, we need to update the txn_summary_num_occurrences only if the batch is already
+            // in author_to_batches and batch_to_proof.
+            let batch_sort_key = BatchSortKey::from_info(&batch_info);
+            let proof_exists_in_author_to_batches = self
+                .author_to_batches
+                .get(&batch_info.author())
+                .map_or(false, |batches| batches.contains_key(&batch_sort_key));
+            let proof_exists_in_batch_to_proof = self
+                .batch_to_proof
+                .get(&batch_sort_key.batch_key)
+                .map_or(false, Option::is_some);
+            let batch_summary_not_already_added = self
+                .batch_summaries
+                .get(&batch_sort_key.batch_key)
+                .map_or(true, |(_, _, added)| !*added);
+            if proof_exists_in_author_to_batches
+                && proof_exists_in_batch_to_proof
+                && batch_summary_not_already_added
             {
-                for txn_summary in txn_summaries {
-                    if let Some(count) = self.txn_summary_num_occurrences.get_mut(&txn_summary) {
-                        *count += 1;
-                    } else {
-                        self.txn_summary_num_occurrences.insert(txn_summary, 1);
-                    }
+                for txn_summary in &txn_summaries {
+                    *self
+                        .txn_summary_num_occurrences
+                        .entry(*txn_summary)
+                        .or_insert(0) += 1;
                 }
             }
+            if batch_summary_not_already_added {
+                self.batch_summaries.insert(
+                    batch_sort_key.batch_key,
+                    (
+                        txn_summaries,
+                        batch_info.expiration(),
+                        proof_exists_in_author_to_batches && proof_exists_in_batch_to_proof,
+                    ),
+                );
+            }
         }
+
+        self.gc_expired_batch_summaries_without_proofs();
```

Comment:
Will this be too expensive to call for every insert? Should we be doing this periodically instead @bchocho ?

Reply:
Here we are iterating through the batches and checking the expiration times. We are not iterating through the transactions. This should be pretty cheap.

Reply:
Still it creates o(n2) operations when n is the number of batches - each insert we iterate through all batches. We should avoid o(n2) operations here.

Reply:
I tested the code with graceful overload test. This is the latency for add batch summaries with the above garbage collection included. 
<img width="636" alt="Screenshot 2024-07-09 at 12 29 13 AM" src="https://github.com/aptos-labs/aptos-core/assets/7092734/22ece7ce-3019-4ebf-b3c3-346f57debbb5">

Reply:
@vusirikala is there a good reason to be doing this on every batch receive? Can we just update on the commit notification?

Reply:
We observed on the testnet that due to some reason, the node failed to receive commit notifications, the node got stuck and the proof queue became large. So, I did the garbage collection here instead.

Reply:
Ah, I see. Yes, this can help with nodes that are lagging on state sync.

I wonder if we can run it less frequently, maybe just log the latest time it was run and only re-run if > 200 ms or so have passed?

Reply:
Sure. I replaced the statement with `sample!` to do the garbage collection once every 500ms.